### DISCUSSION TITLE
debug.sh 显示结果里有223.5.5.5这类公共dns记录，另外dig的query time按理说是0ms，但是一直是跟阿里dns查询时间接近。

### DISCUSSION THREAD

**Discussion by tonebean** [kkkgo/PaoPaoDNS#64] [问答和帮助]
/data # debug.sh
=====PaoPaoDNS docker debug=====
[info] images build time : 2023-11-09 16:17:17 UTC
[DEBUG-OK]DATA_writeable
[DEBUG-OK]DATA_readable
[info] ========== network info ==========
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
13: eth0@if2: <BROADCAST,MULTICAST,UP,LOWER_UP,M-DOWN> mtu 1500 qdisc noqueue state UP
    link/ether 02:42:c0:a8:00:dc brd ff:ff:ff:ff:ff:ff
    inet 192.168.0.220/24 brd 192.168.0.255 scope global eth0
       valid_lft forever preferred_lft forever
default via 192.168.0.1 dev eth0
192.168.0.0/24 dev eth0 scope link  src 192.168.0.220
PING 223.5.5.5 (223.5.5.5): 56 data bytes
64 bytes from 223.5.5.5: seq=0 ttl=116 time=6.286 ms
64 bytes from 223.5.5.5: seq=1 ttl=116 time=6.100 ms

--- 223.5.5.5 ping statistics ---
2 packets transmitted, 2 packets received, 0% packet loss
round-trip min/avg/max = 6.100/6.193/6.286 ms
PING 119.29.29.29 (119.29.29.29): 56 data bytes
64 bytes from 119.29.29.29: seq=0 ttl=52 time=36.915 ms
64 bytes from 119.29.29.29: seq=1 ttl=52 time=36.754 ms

--- 119.29.29.29 ping statistics ---
2 packets transmitted, 2 packets received, 0% packet loss
round-trip min/avg/max = 36.754/36.834/36.915 ms
Server:         223.5.5.5
Address:        223.5.5.5#53

Non-authoritative answer:
www.taobao.com  canonical name = www.taobao.com.danuoyi.tbcache.com.
Name:   www.taobao.com.danuoyi.tbcache.com
Address: 118.112.23.199
Name:   www.taobao.com.danuoyi.tbcache.com
Address: 118.112.23.198
Name:   www.taobao.com.danuoyi.tbcache.com
Address: 240e:932:c001:3:3::3da
Name:   www.taobao.com.danuoyi.tbcache.com
Address: 240e:932:c001:3:3::3db

Server:         119.29.29.29
Address:        119.29.29.29#53

Non-authoritative answer:
www.qq.com      canonical name = ins-r23tsuuf.ias.tencent-cloud.net.
Name:   ins-r23tsuuf.ias.tencent-cloud.net
Address: 121.14.77.221
Name:   ins-r23tsuuf.ias.tencent-cloud.net
Address: 121.14.77.201
Name:   ins-r23tsuuf.ias.tencent-cloud.net
Address: 240e:97c:2f:2::4c
Name:   ins-r23tsuuf.ias.tencent-cloud.net
Address: 240e:97c:2f:1::5c

[info] ========== env info ==========
====ENV TEST====
[OK]DATA_writeable-
[OK]DATA_readable-
MEM:400m 800m 500000 900mb
prefPC:82
CORES:-2-
POWCORES:-2-
ulimit :-1048576-
FDLIM :-4096-
TZ:-Asia/Shanghai-
UPDATE:-weekly-
DNS_SERVERNAME:-PaoPaoDNS-
SERVER_IP:-none-
ETHIP:-192.168.0.220-
DNSPORT:-53-
SOCKS5:-IP:PORT-
CNAUTO:-yes-
IPV6:-no-
CNFALL:-yes-
CUSTOM_FORWARD:-IP:PORT-
AUTO_FORWARD:-no-
AUTO_FORWARD_CHECK:-yes-
USE_MARK_DATA:-yes-
RULES_TTL:-0-
CUSTOM_FORWARD_TTL:-0-
SHUFFLE:-no-
CN_TRACKER:-yes-
USE_HOSTS:-no-
HTTP_FILE:-no-
SAFEMODE:--
QUERY_TIME:-2000ms-
ADDINFO:-yes-
PLATFORM:-Linux e42c19596aeb 6.1.0-13-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.1.55-1 (2023-09-29) x86_64 Linux-
====ENV TEST====
mosdns kkkgo/mosdns:231106.1
[info] ========== process info ==========
PID   USER     TIME  COMMAND
    1 root      0:00 {init.sh} /bin/sh /usr/sbin/init.sh
   13 root      0:00 crond
   44 root      0:04 redis-server unixsocket:/tmp/redis.sock
  126 root      0:02 dnscrypt-proxy -config /data/dnscrypt-resolvers/dnscrypt.toml
  137 root      0:03 mosdns start -d /tmp -c /tmp/mosdns.yaml
  145 root      0:00 unbound -c /tmp/unbound_forward.conf -p
  154 root      0:00 {watch_list.sh} /bin/sh /usr/sbin/watch_list.sh
  161 root      0:00 tail -f /dev/null
  183 root      0:05 unbound -c /tmp/unbound_raw.conf -p
  193 root      0:00 inotifywait -e modify,delete /etc/unbound/named.cache /data/Country-only-cn-private.mmdb /data/force_cn_list.txt /data/force_nocn_list.txt /data/custom_env.ini /data/global_mark.dat /data/trackerslist
  679 root      0:00 sh
  686 root      0:00 {debug.sh} /bin/sh /usr/sbin/debug.sh
  702 root      0:00 ps -ef
[info] ========== cn list info ==========
domain:whoami.ds.akahelp.net
domain:whoami.03k.org
[info] ========== reids info ==========
used_memory_human:1.52M
used_memory_rss_human:4.82M
used_memory_peak_human:1.53M
total_system_memory_human:5.79G
used_memory_lua_human:31.00K
used_memory_vm_total_human:63.00K
used_memory_scripts_human:181B
maxmemory_human:900.00M
(integer) 1190
[test] IP test, you will see that all the following IPs are your public network exit IP !
[test] ========== IP TEST START ==========
[test] ========== IP TEST END ==========
[test] The DNS hijacking test, you will see timed out message !
[test] ========== DNS HIJACK START ==========
;; communications error to 9.8.7.6#53: timed out
;; communications error to 9.8.7.6#53: timed out
;; communications error to 9.8.7.6#53: timed out
;; no servers could be reached

[test] ========== DNS HIJACK END ==========
[test] CN domain test, you will see that the DNS resolution result is CN IP !
[test] ========== CN DOMAIN TEST START ==========
----mosdns CN dig:
www.taobao.com.danuoyi.tbcache.com.
118.112.23.199
118.112.23.198
------------------
----local-unbound CN dig:
www.taobao.com.danuoyi.tbcache.com.
118.112.23.199
118.112.23.198
[test] ========== CN DOMAIN TEST END ==========
[test] Non-CN domain test, you will see that the DNS resolution result is correct IP !
[test] ========== Non-CN DOMAIN TEST START ==========
----mosdns Non-CN dig:
youtube-ui.l.google.com.
142.251.32.46
142.251.214.142
------------------
----dnscrypt-unbound NOCN dig:
youtube-ui.l.google.com.
142.250.80.78
142.250.80.110
------------------
----dnscrypt Non-CN dig:
youtube-ui.l.google.com.
142.250.179.238
142.250.180.14
------------------
----dnscrypt-socks5 Non-CN dig:
;; communications error to 127.0.0.1#5303: connection refused
;; communications error to 127.0.0.1#5303: connection refused
;; communications error to 127.0.0.1#5303: connection refused
[test] ========== Non-CN DOMAIN TEST END ==========
[test] IPv6 Dual CN test: you will see that IPv6 is OK !
[test] ========== IPV6 CN DOMAIN TEST START ==========
[test] ========== IPV6 CN DOMAIN TEST END ==========
[test] IPv6 Dual Non-CN test: you will see that IPv6 is empty !
[test] ========== IPV6 Non-CN DOMAIN TEST START ==========
[test] ========== IPV6 Non-CN DOMAIN TEST END ==========
[test] IPv6 only Non-CN test: you will see that IPv6 is ok !
[test] ========== IPV6 ONLY Non-CN DOMAIN TEST START ==========
[test] ========== IPV6 ONLY Non-CN DOMAIN TEST END ==========

[info] ALL TEST FINISH.

**Comment by kkkgo** [comment]
debug.sh是用223.5.5.5测试网络环境有没有问题。

**Reply by tonebean** [reply]
多谢回复，现在唯一的问题是dig国内域名，重复查询的时间都是接近阿里dns查询，怀疑是cnfall的问题，按理说有了cache，query time应该是0。

**Reply by kkkgo** [reply]
你可以开addinfo来观察结果来源。

**Reply by tonebean** [reply]
第n次查询
; <<>> DiG 9.10.6 <<>> baidu.com @192.168.0.250 -p 5353
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 22102
;; flags: qr rd ra; QUERY: 1, ANSWER: 2, AUTHORITY: 0, ADDITIONAL: 2

;; QUESTION SECTION:
;baidu.com.			IN	A

;; ANSWER SECTION:
baidu.com.		311	IN	A	110.242.68.66
baidu.com.		311	IN	A	39.156.66.10

;; ADDITIONAL SECTION:
20231110160154.634484096.addinfo.paopaodns. 312	IN TXT "Since 10ms From:mmdb CN IP"
20231110160345.757357667.addinfo.paopaodns. 312	IN TXT "Since 0ms From:respond mosdns cache"

;; Query time: 6 msec
;; SERVER: 192.168.0.250#5353(192.168.0.250)
;; WHEN: Fri Nov 10 15:46:45 CST 2023
;; MSG SIZE  rcvd: 248

第n+1次查询
; <<>> DiG 9.10.6 <<>> baidu.com @192.168.0.250 -p 5353
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 31704
;; flags: qr rd ra; QUERY: 1, ANSWER: 2, AUTHORITY: 0, ADDITIONAL: 2

;; QUESTION SECTION:
;baidu.com.			IN	A

;; ANSWER SECTION:
baidu.com.		311	IN	A	110.242.68.66
baidu.com.		311	IN	A	39.156.66.10

;; ADDITIONAL SECTION:
20231110160154.634484096.addinfo.paopaodns. 312	IN TXT "Since 10ms From:mmdb CN IP"
20231110160346.100471869.addinfo.paopaodns. 312	IN TXT "Since 0ms From:respond mosdns cache"

;; Query time: 12 msec
;; SERVER: 192.168.0.250#5353(192.168.0.250)
;; WHEN: Fri Nov 10 15:46:45 CST 2023
;; MSG SIZE  rcvd: 248


很奇怪，addinfo的cache是0ms，结果不管dig多少次这个时间都不稳，理论上说Query time应该趋向于0ms。

**Reply by kkkgo** [reply]
`Since 0ms From:respond mosdns cache`说明查询实际内部耗时是0ms。12ms是DNS服务器到你客户端的网络延迟。

**Reply by tonebean** [reply]
谢谢解答，没想到家庭网络内部有这么长的延迟，我一直以为是1ms这样子。

**Reply by kkkgo** [reply]
另外，`"Since 10ms From:mmdb CN IP"`是表示历史递归所需时间。你可以看到第n和n+1次该信息的时间戳是一样的。
DNS服务器和客户端网络延迟可能与你的硬件资源设备，以及容器网络模式有关系。

**Reply by kkkgo** [reply]
你可以尝试macvlan网络或者host网络，以及结合ping来观察延迟。

**Reply by tonebean** [reply]
macvlan和host都试过，情况一样，ping是<1ms。

**Reply by kkkgo** [reply]
你可以分别在容器内部和容器外部dig来看延迟。

**Reply by tonebean** [reply]
应该是我内部网络的问题，dig  tun网关 fakeip也是那个延迟时间。
